File: packages/distance/src/edit/ratcliff.ts
/**
 * Ratcliff-Obershelp algorithm — Gestalt pattern matching.
 *
 * Iteratively finds the longest common substring using a stack-based approach,
 * combining scores from both sides. Returns a similarity in [0, 1].
 *
 * Based on the textdistance.rs implementation.
 *
 * Time: O(n * m * log(n * m)) worst case, O(n + m) average
 */

/**
 * Internal: find the longest common substring and return its length and positions.
 */
function findLCS(a: string, b: string): { len: number; aIdx: number; bIdx: number } {
  const aLen = a.length;
  const bLen = b.length;

  if (aLen === 0 || bLen === 0) return { len: 0, aIdx: 0, bIdx: 0 };

  let maxLen = 0;
  let endI = 0;
  let endJ = 0;
  const dp = new Uint32Array(bLen + 1);

  for (let i = 1; i <= aLen; i++) {
    let prev = 0;
    for (let j = 1; j <= bLen; j++) {
      const temp = dp[j];
      if (a.charCodeAt(i - 1) === b.charCodeAt(j - 1)) {
        dp[j] = prev + 1;
        if (dp[j] > maxLen) {
          maxLen = dp[j];
          endI = i;
          endJ = j;
        }
      } else {
        dp[j] = 0;
      }
      prev = temp;
    }
  }

  return { len: maxLen, aIdx: endI - maxLen, bIdx: endJ - maxLen };
}

/**
 * Compute Ratcliff-Obershelp similarity between two strings.
 *
 * Uses an iterative stack-based approach to avoid stack overflow on
 * very different strings. The algorithm recursively finds the longest
 * common substring and combines similarity scores from both sides.
 *
 * similarity = 2 * M / T, where M = total matched characters, T = total characters
 *
 * @param a - First string
 * @param b - Second string
 * @returns Ratcliff-Obershelp similarity in [0, 1]
 */
export function ratcliff(a: string, b: string): number {
  if (a === b) return 1;
  const totalLen = a.length + b.length;
  if (totalLen === 0) return 1;

  let totalMatch = 0;

  // Stack of (aLeft, aRight, bLeft, bRight) pairs to process
  const stack: [number, number, number, number][] = [[0, a.length, 0, b.length]];

  while (stack.length > 0) {
    const [aStart, aEnd, bStart, bEnd] = stack.pop()!;

    if (aEnd - aStart === 0 || bEnd - bStart === 0) continue;

    const subA = a.slice(aStart, aEnd);
    const subB = b.slice(bStart, bEnd);

    const lcs = findLCS(subA, subB);

    if (lcs.len === 0) continue;

    totalMatch += lcs.len;

    // Push right sides first (processed after left due to LIFO)
    const aRightStart = aStart + lcs.aIdx + lcs.len;
    const bRightStart = bStart + lcs.bIdx + lcs.len;
    if (aEnd - aRightStart > 0 && bEnd - bRightStart > 0) {
      stack.push([aRightStart, aEnd, bRightStart, bEnd]);
    }

    // Push left sides
    const aLeftEnd = aStart + lcs.aIdx;
    const bLeftEnd = bStart + lcs.bIdx;
    if (aLeftEnd - aStart > 0 && bLeftEnd - bStart > 0) {
      stack.push([aStart, aLeftEnd, bStart, bLeftEnd]);
    }
  }

  return (2 * totalMatch) / totalLen;
}

File: packages/distance-wasm/src/edit/ratcliff.rs
/// Ratcliff-Obershelp — Gestalt pattern matching.
///
/// Iteratively finds longest common substrings using a stack-based approach.
///
/// Time: O(n * m * k) where k = recursion depth

pub fn ratcliff_obershelp(a: &str, b: &str) -> f64 {
    if a == b { return 1.0; }
    let total_len = a.len() + b.len();
    if total_len == 0 { return 1.0; }

    let a_bytes = a.as_bytes();
    let b_bytes = b.as_bytes();
    let mut total_match = 0u32;

    // Stack of (a_start, a_end, b_start, b_end)
    let mut stack: Vec<(usize, usize, usize, usize)> = vec![(0, a.len(), 0, b.len())];

    while let Some((a_start, a_end, b_start, b_end)) = stack.pop() {
        if a_end == a_start || b_end == b_start { continue; }

        let lcs = find_lcs(&a_bytes[a_start..a_end], &b_bytes[b_start..b_end]);
        if lcs.len == 0 { continue; }

        total_match += lcs.len as u32;

        // Push right side
        let a_right = a_start + lcs.a_idx + lcs.len;
        let b_right = b_start + lcs.b_idx + lcs.len;
        if a_end > a_right && b_end > b_right {
            stack.push((a_right, a_end, b_right, b_end));
        }

        // Push left side
        if lcs.a_idx > 0 && lcs.b_idx > 0 {
            stack.push((a_start, a_start + lcs.a_idx, b_start, b_start + lcs.b_idx));
        }
    }

    2.0 * total_match as f64 / total_len as f64
}

struct LcsResult {
    len: usize,
    a_idx: usize,
    b_idx: usize,
}

/// Find the longest common substring and its positions.
fn find_lcs(a: &[u8], b: &[u8]) -> LcsResult {
    let a_len = a.len();
    let b_len = b.len();

    if a_len == 0 || b_len == 0 {
        return LcsResult { len: 0, a_idx: 0, b_idx: 0 };
    }

    let mut max_len = 0usize;
    let mut end_i = 0usize;
    let mut end_j = 0usize;
    let mut dp = vec![0u32; b_len + 1];

    for i in 1..=a_len {
        let mut prev = 0u32;
        for j in 1..=b_len {
            let temp = dp[j];
            if a[i - 1] == b[j - 1] {
                dp[j] = prev + 1;
                if dp[j] as usize > max_len {
                    max_len = dp[j] as usize;
                    end_i = i;
                    end_j = j;
                }
            } else {
                dp[j] = 0;
            }
            prev = temp;
        }
    }

    LcsResult {
        len: max_len,
        a_idx: end_i - max_len,
        b_idx: end_j - max_len,
    }
}
